images/virtualization-artifact/pkg/controller/vm/internal/provisioning.go

```diff
@@ -120,30 +118,106 @@ func (h *ProvisioningHandler) Name() string {
 	return nameProvisioningHandler
 }
 
-func (h *ProvisioningHandler) genConditionFromSecret(ctx context.Context, builder *conditions.ConditionBuilder, secretKey types.NamespacedName, checkKeys ...string) error {
-	secret, err := helper.FetchObject(ctx, secretKey, h.client, &corev1.Secret{})
-	if err != nil {
-		return fmt.Errorf("failed to fetch secret: %w", err)
-	}
-	if secret == nil {
+func (h *ProvisioningHandler) genConditionFromSecret(ctx context.Context, builder *conditions.ConditionBuilder, secretKey types.NamespacedName) error {
+	var (
+		errUnexpectedSecretType unexpectedSecretTypeError
+		errSecretNotFound       secretNotFoundError
+	)
+	err := h.validator.Validate(ctx, secretKey)
+
+	switch {
+	case err == nil:
+		builder.Reason(vmcondition.ReasonProvisioningReady).Status(metav1.ConditionTrue)
+		return nil
+	case errors.As(err, &errSecretNotFound):
```

Comment:
To avoid using additional variable errSecretNotFound:

`case errors.As(err, new(secretNotFoundError))`

And same for errUnexpectedSecretType